changed.d

```diff
@@ -142,56 +157,231 @@ string getChangeLog(string revRange)
         {
-            result ~= format("$(BUGSTITLE %s %s,\n\n", component, bugtype);
-            foreach (bug; sort!"a.id < b.id"(*bugs))
+            foreach (bugtype; bugtypes)
+            if (auto bugs = bugtype in *comp)
             {
-                result ~= format("$(LI $(BUGZILLA %s): %s)\n",
-                              bug.id, bug.summary.escapeParens());
+                w.formattedWrite("$(BUGSTITLE %s %s,\n\n", component, bugtype);
+                foreach (bug; sort!"a.id < b.id"(*bugs))
+                {
+                    w.formattedWrite("$(LI $(BUGZILLA %s): %s)\n",
+                                        bug.id, bug.summary.escapeParens());
+                }
+                w.put(")\n");
             }
-            result ~= ")\n";
         }
     }
-
-    return result.data;
 }
 
 int main(string[] args)
 {
-    if (args.length != 2)
+    auto outputFile = "./changelog.dd";
+    auto nextVersionString = "LATEST";
+    auto nextVersionDate = "September 19, 2016";
+    string previousVersion = "Previous version";
+    bool hideTextChanges = false;
+    string revRange;
+
+    auto helpInformation = getopt(
+        args,
+        std.getopt.config.passThrough,
+        "output|o", &outputFile,
+        "date", &nextVersionDate,
+        "version", &nextVersionString,
+        "prev-version", &previousVersion, // this can automatically be detected
+        "no-text", &hideTextChanges);
+
+    if (args.length >= 2)
     {
-        stderr.writeln("Usage: ./changed <revision range>, e.g. ./changed v2.067.1..upstream/stable");
-        return 1;
+        revRange = args[1];
+
+        // extract the previous version
+        auto parts = revRange.split("..");
+        if (parts.length > 1)
+            previousVersion = parts[0];
     }
+    else
+    {
+        writeln("Skipped querying Bugzilla for changes. Please define a revision range e.g ./changed v2.072.2..upstream/stable");
+    }
+
+    auto f = File(outputFile, "w");
+    auto w = f.lockingTextWriter();
+    w.put("Ddoc\n\n");
+    w.formattedWrite("$(CHANGELOG_NAV_LAST %s)\n\n", previousVersion);
+
+    {
+        w.formattedWrite("$(VERSION %s, =================================================,\n\n", nextVersionDate);
+        scope(exit) w.put(")\n");
+
+        if (!hideTextChanges)
+        {
+            // search for raw change files
+            alias Repo = Tuple!(string, "path", string, "headline");
+            auto repos = [Repo("dmd", "Language changes"),
+                          Repo("druntime", "Runtime changes"),
+                          Repo("phobos", "Library changes")];
+
+            auto changedRepos = repos
+                 .map!(repo => Repo(buildPath("..", repo.path, "changelog"), repo.headline))
+                 .filter!(x => x.path.exists)
+                 .map!(x => tuple!("headline", "changes")(x.headline, x.path.readTextChanges.array))
+                 .filter!(x => !x.changes.empty);
```

Comment:
Please keep a literal and understandable lambda argument name instead of using `x`, `r` for repo would be fine.
Can live w/ the x, but missing names are usually a code smell, b/c it requires the code reader to extract a mental model from raw code, instead of having the writer present one.